### PR TITLE
[Move examples] Add an example for locked coins contract

### PR DESCRIPTION
**Pull request by movekevin**
### Description
Simple module that can be used to lock up coins until an unlock timestamp

### Test Plan
Unit tests

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/5334)

### CODE REVIEW

File: aptos-move/framework/cached-packages/src/aptos_framework_sdk_builder.rs [hunk 1]
```diff
@@ -144,6 +144,24 @@ pub enum EntryFunctionCall {
         coin_type: TypeTag,
     },
 
+    /// `Sponsor` can add locked coins for `recipient` with given unlock timestamp (in seconds).
+    /// There's no restriction on unlock timestamp so sponsors could technically add coins for an unlocked time in the
+    /// past, which means the coins are immediately unlocked.
+    LockedCoinsAddLockedCoins {
```
Comment (areshand) [inline aptos-move/framework/cached-packages/src/aptos_framework_sdk_builder.rs:150]: Do we want framework SDK generating function for move-example now?

File: aptos-move/move-examples/defi/sources/locked_coins.move [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    use aptos_std::table::{Self, Table};
+    use std::error;
+    use std::signer;
+
+    /// Represents a lock of coins until some specified unlock time. Afterward, the recipient can claim the coins.
+    struct Lock<phantom CoinType> has store {
+        coins: Coin<CoinType>,
+        unlock_time_secs: u64,
+    }
+
```
Comment (chloeqjz) [inline aptos-move/move-examples/defi/sources/locked_coins.move:15]: I get that the locks table is `<recipient_address, locked_coins of CoinType for this recipient>` but it'd be nice to have a comment making it clearer : )

File: aptos-move/move-examples/defi/sources/locked_coins.move [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            move_to(sponsor, Locks {
+                locks: table::new<address, Lock<CoinType>>(),
+                claim_events: account::new_event_handle<ClaimEvent>(sponsor),
+            })
+        };
+
+        let locks = borrow_global_mut<Locks<CoinType>>(sponsor_address);
+        let coins = coin::withdraw<CoinType>(sponsor, amount);
+        if (table::contains(&locks.locks, recipient)) {
+            let lock = table::borrow_mut(&mut locks.locks, recipient);
```
[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/move-examples/defi/sources/locked_coins.move:49]: That's a good callout. Adding more coins to an existing lock is not a critical use case so may be we should just explicitly omit it for simplicity.

[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/move-examples/defi/sources/locked_coins.move]: Updated

File: aptos-move/move-examples/defi/sources/locked_coins.move (hunk 2 above)
Comment (movekevin) [reply aptos-move/move-examples/defi/sources/locked_coins.move:15]: Updated

File: aptos-move/framework/cached-packages/src/aptos_framework_sdk_builder.rs (hunk 1 above)
Comment (movekevin) [reply aptos-move/framework/cached-packages/src/aptos_framework_sdk_builder.rs:150]: This seems to have always been here. If you think we should remove, we can discuss this separately :)

**Review by chloeqjz — APPROVED** [review]

File: aptos-move/move-examples/defi/sources/locked_coins.move [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    /// Represents a lock of coins until some specified unlock time. Afterward, the recipient can claim the coins.
+    struct Lock<phantom CoinType> has store {
+        coins: Coin<CoinType>,
+        unlock_time_secs: u64,
+    }
+
+    /// Holder for a map from recipients => locks.
+    /// There can be at most one lock per recipient.
+    struct Locks<phantom CoinType> has key {
+        locks: Table<address, Lock<CoinType>>,
```
Comment (areshand) [inline aptos-move/move-examples/defi/sources/locked_coins.move:19]: what happens if different amount of coined with different lock time per address?

File: aptos-move/move-examples/defi/sources/locked_coins.move [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            move_to(sponsor, Locks {
+                locks: table::new<address, Lock<CoinType>>(),
+                claim_events: account::new_event_handle<ClaimEvent>(sponsor),
+            })
+        };
+
+        let locks = borrow_global_mut<Locks<CoinType>>(sponsor_address);
+        let coins = coin::withdraw<CoinType>(sponsor, amount);
+        assert!(!table::contains(&locks.locks, recipient), error::already_exists(ELOCK_ALREADY_EXISTS));
+        table::add(&mut locks.locks, recipient, Lock<CoinType> { coins, unlock_time_secs });
```
Comment (areshand) [inline aptos-move/move-examples/defi/sources/locked_coins.move:53]: if no one is claiming the coin, this coin will be locked forever. should we provide option that claimer has to claim the coin within a time framework. beyond a time, the sponsor can re-claim the offered coin

File: aptos-move/move-examples/defi/sources/locked_coins.move (hunk 4 above)
Comment (movekevin) [reply aptos-move/move-examples/defi/sources/locked_coins.move:19]: That's not a problem. In fact, it's how this is meant to work :) You can give different amounts with different lock durations to different accounts

File: aptos-move/move-examples/defi/sources/locked_coins.move (hunk 5 above)
Comment (movekevin) [reply aptos-move/move-examples/defi/sources/locked_coins.move:53]: Good point regarding expiration. It's an interesting feature though 🤔 For now I'd keep it simple and not have that feature, but we can certainly revisit and include it in the future if the community wants to bring this module into the Aptos framework.

**Review by areshand — APPROVED** [review]

**Pull request merged**